### PR TITLE
Accept iterables.

### LINKED ISSUE #67: Iteration methods such as node.eachBefore should pass the index, too.
I'm upgrading some code from v3 to v4. I think it'd be useful for `eachBefore` and perhaps the other loops to receive an index. e.g. Migrating this block: http://bl.ocks.org/mbostock/1093025

### LINKED ISSUE #94: node.find?
It would be useful to have a `node#find` function.

In addition, in Lodash, returning `false` in a `forEach` callback exits early. This would be useful in `node#each` for performance optimisations for custom iteration functions.

Happy to submit a PR if you want to go in this direction.

### LINKED ISSUE #140: Accept iterables as children.
It’d be nice to make d3.hierarchy compatible with the new d3.group in d3-array.

### LINKED ISSUE #154: Support iterables.
- [x] d3.hierarchy(…) *children* #140 
- [x] d3.stratify(*data*)
- [x] d3.packSiblings(*circles*)
- [x] d3.packEnclose(*circles*)

### PR DESCRIPTION
**Pull request by mbostock**
Fixes #94.
Fixes #140.
Fixes #154.
Fixes #67.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -174,17 +174,27 @@ root
 
 You must call *node*.sort before invoking a hierarchical layout if you want the new sort order to affect the layout; see [*node*.sum](#node_sum) for an example.
 
-<a name="node_each" href="#node_each">#</a> <i>node</i>.<b>each</b>(<i>function</i>) · [Source](https://github.com/d3/d3-hierarchy/blob/master/src/hierarchy/each.js), [Examples](https://observablehq.com/@d3/visiting-a-d3-hierarchy)
+<a name="node_iterator" href="#node_iterator">#</a> <i>node</i>[<b>Symbol.iterator</b>]() [<>](https://github.com/d3/d3-hierarchy/blob/master/src/hierarchy/iterator.js "Source")
```
Comment (Fil) [inline README.md]: this looks like the old format for links to Source/Examples.

File: test/hierarchy/index-test.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
+var tape = require("tape"),
+    d3 = require("../../");
+
+tape("d3.hierarchy(data, children) supports iterable children", function(test) {
```
Comment (Fil) [inline test/hierarchy/index-test.js]: a detail, but the "children" argument listed in this comment is not used

**Comment by mbostock** [comment]
Sure, I’d love test contributions.

**Comment by mbostock** [comment]
I made some tweaks so that you can pass a rollup or group Map directly to d3.hierarchy and it’ll work out of the box, and I’ve been tinkering with the d3-hierarchy notebooks to make it easier to override the data (e.g., using d3.stratify or d3.rollup without needing to change the rest of the code). Let me know if you have any thoughts there.

**Comment by mbostock** [comment]
One joyful discovery is that, not that hierarchies are iterable, they work beautifully with the new d3-array, e.g.

```js
const [x0, x1] = d3.extent(root, d => d.x);
```

**Comment by Fil** [comment]
This example is a bit frustrating to port to v6
https://observablehq.com/d/d72463e9ecde64e6

**Pull request merged**